### PR TITLE
Name resolution globbing

### PR DESCRIPTION
**Pull request by P-E-P**
Late name resolution was not used and did not handle globbing.

### CODE REVIEW

**Comment by powerboat9** [comment]
@CohenArthur how much of this can be merged as-is?

**Comment by P-E-P** [comment]
I still need to reorder some commits and fix the changelog but this is nearly done :tada:

File: gcc/rust/resolve/rust-toplevel-name-resolver-2.0.h [hunk 1]
```diff
@@ -59,8 +82,10 @@ class TopLevel : public DefaultResolver
 
   // FIXME: Do we move these to our mappings?
   std::unordered_map<NodeId, location_t> node_locations;
+  std::unordered_map<NodeId, NodeId> node_forwarding;
```
Comment (CohenArthur) [inline gcc/rust/resolve/rust-toplevel-name-resolver-2.0.h:85]: an you add a small comment explaining the point of this map?

**Review by CohenArthur — COMMENTED** [review]
I don't think there are any test cases for glob imports, could we add some?

File: gcc/rust/resolve/rust-toplevel-name-resolver-2.0.cc [hunk 2]
```diff
@@ -298,20 +425,45 @@ TopLevel::visit (AST::Union &union_item)
...
 {
   // TODO: Glob imports can get shadowed by regular imports and regular items.
   // So we need to store them in a specific way in the ForeverStack - which can
   // also probably be used by labels and macros etc. Like store it as a
   // `Shadowable(NodeId)` instead of just a `NodeId`
+  auto resolved = ctx.types.resolve_path (glob.get_segments ());
+  if (!resolved.has_value ())
+    {
+      return false;
+    }
```
[the remark this answers is not in the document]
Comment (P-E-P) [reply gcc/rust/resolve/rust-toplevel-name-resolver-2.0.cc]: Fixed.

File: gcc/rust/backend/rust-compile-item.cc [hunk 3]
```diff
@@ -149,12 +166,32 @@ CompileItem::visit (HIR::Function &function)
...
-  bool ok = ctx->get_mappings ()->lookup_canonical_path (
-    function.get_mappings ().get_nodeid (), &canonical_path);
-  rust_assert (ok);
+  Resolver::CanonicalPath canonical_path
+    = Resolver::CanonicalPath::create_empty ();
+
+  if (flag_name_resolution_2_0)
+    {
+      // FIXME: What we need to do here is build the CanonicalPath from the name
+      // resolver 2.0. this isn't super difficult, we just need to get to it
```
[the remark this answers is not in the document]
Comment (P-E-P) [reply gcc/rust/backend/rust-compile-item.cc]: Fixed.

File: gcc/rust/resolve/rust-default-resolver.cc [hunk 4]
```diff
@@ -79,8 +79,13 @@ DefaultResolver::visit (AST::Function &function)
 	  }
       }
 
+    if (function.has_return_type ())
+      visit (function.get_return_type ());
+
     if (function.has_body ())
       function.get_definition ().value ()->accept_vis (*this);
+
+    // TODO: Don't we need to visit the return type here as well?
```
[the remark this answers is not in the document]
Comment (P-E-P) [reply gcc/rust/resolve/rust-default-resolver.cc]: Fixed.

File: gcc/rust/resolve/rust-toplevel-name-resolver-2.0.cc [hunk 5]
```diff
@@ -437,44 +719,67 @@ flatten_list (const AST::UseTreeList &list, std::vector<AST::SimplePath> &paths)
...
+  // 1. figure out the rib which contains all the exports, in each namespace
+  // 2. for that namespace, go through all the nodes
+  // 3. filter out the exported ones (is that correct?)
+  // 4. export them in the current rib, BUT make them shadowable
+
+  // 1. needs us to be able to find a Rib based on a SimplePath - can we already
+  // do that with resolve path?
+  // 2. ok
+  // 3. how do we get the visibiliy information for each node? in the mappings?
+  // 4. ok
```
[the remark this answers is not in the document]
Comment (P-E-P) [reply gcc/rust/resolve/rust-toplevel-name-resolver-2.0.cc]: > so do we need to do anything for flatten_glob?

Flatten glob only need to push the current path into the list. 

> can we remove these comments?

Fixed.

File: gcc/rust/resolve/rust-toplevel-name-resolver-2.0.cc [hunk 6]
```diff
@@ -322,38 +474,137 @@ TopLevel::handle_use_dec (AST::SimplePath path)
...
+	      && result->second != path.get_node_id ())
+	    rust_error_at (path.get_locus (), "%qs defined multiple times",
+			   declared_name.c_str ());
+	}
+      else
+	{
+	  // No previous thing has inserted this into our scope
+	  node_forwarding.insert ({def.get_node_id (), path.get_node_id ()});
+	}
+      return def.get_node_id ();
```
[the remark this answers is not in the document]
Comment (P-E-P) [reply gcc/rust/resolve/rust-toplevel-name-resolver-2.0.cc:599]: No they're not exactly the same. One need to handle the aliased name.

File: gcc/rust/resolve/rust-toplevel-name-resolver-2.0.h (hunk 1 above)
Comment (P-E-P) [reply gcc/rust/resolve/rust-toplevel-name-resolver-2.0.h:88]: Fixed.

File: gcc/rust/resolve/rust-toplevel-name-resolver-2.0.cc [hunk 7]
```diff
@@ -25,6 +26,105 @@
...
+void
+GlobbingVisitor::visit (AST::TupleStruct &tuple_struct)
+{
+  if (tuple_struct.get_visibility ().is_public ())
+    {
+      ctx.insert_shadowable (tuple_struct.get_identifier (),
+			     tuple_struct.get_node_id (), Namespace::Types);
+
+      ctx.insert_shadowable (tuple_struct.get_identifier (),
+			     tuple_struct.get_node_id (), Namespace::Values);
```
[the remark this answers is not in the document]
Comment (P-E-P) [reply gcc/rust/resolve/rust-toplevel-name-resolver-2.0.cc:90]: No, we need to insert all constructors.

File: gcc/rust/resolve/rust-toplevel-name-resolver-2.0.h [hunk 8]
```diff
@@ -60,7 +83,12 @@ class TopLevel : public DefaultResolver
   // FIXME: Do we move these to our mappings?
   std::unordered_map<NodeId, location_t> node_locations;
 
+  // Store node forwarding for use declaration, the link between a
+  // "new" local name and it's definition.
```
Comment (CohenArthur) [inline gcc/rust/resolve/rust-toplevel-name-resolver-2.0.h:87]:
```suggestion
  // "new" local name and its definition.
```

**Review by CohenArthur — APPROVED** [review]
LGTM :) amazing work, thank you for spending so much time on it

File: gcc/rust/resolve/rust-toplevel-name-resolver-2.0.h (hunk 8 above)
Comment (P-E-P) [reply gcc/rust/resolve/rust-toplevel-name-resolver-2.0.h]: Fixed.

**Comment by P-E-P** [comment]
> I don't think there are any test cases for glob imports, could we add some?

I've added two tests, `name_resolution23.rs` and `name_resolution24.rs`. The first one checks the globbing works and ambiguity errors are not triggered when the ambiguous symbol is not used. The second one ensure an error is emitted when an ambiguous symbol is called.

**Pull request merged**